src/Microsoft.Diagnostics.EventFlow.Outputs.EventHub/EventHubOutput.cs

```diff
@@ -83,52 +84,51 @@ public EventHubOutput(IConfiguration configuration, IHealthReporter healthReport
 
             try
             {
-                // Since event hub limits each message/batch to be a certain size, we need to
-                // keep checking the size for exceeds and split into a new batch as needed
+                var groupedEventData = events.GroupBy(e => string.IsNullOrEmpty(outputConfiguration.PartitionKeyProperty) == false && e.TryGetPropertyValue(outputConfiguration.PartitionKeyProperty, out var partionKeyData) ? partionKeyData.ToString() : string.Empty, e => e);
 
-                List<List<MessagingEventData>> batches = new List<List<MessagingEventData>>();
-                int batchByteSize = 0;
+                List<Task> tasks = new List<Task>();
 
-                foreach (EventData eventData in events)
+                foreach (var partitionedEventData in groupedEventData)
                 {
-                    int messageSize;
-                    MessagingEventData messagingEventData = eventData.ToMessagingEventData(out messageSize);
+                    SendBatch(partitionedEventData.Select(e => new Tuple<MessagingEventData, int>(e.ToMessagingEventData(out var messageSize), messageSize)).ToList());
 
-                    // If we don't have a batch yet, or the addition of this message will exceed the limit for this batch, then
-                    // start a new batch.
-                    if (batches.Count == 0 ||
-                        batchByteSize + messageSize > EventHubMessageSizeLimit)
+                    void SendBatch(List<Tuple<MessagingEventData, int>> batch)
                     {
-                        batches.Add(new List<MessagingEventData>());
-                        batchByteSize = 0;
+                        // Since event hub limits each message/batch to be a certain size, we need to
+                        // keep checking the size in bytes of the batch and recursively keep splitting into two batches as needed
+                        if (batch.Count >= 2 && batch.Sum(b => b.Item2) > EventHubMessageSizeLimit)
```

Comment:
Noo :-) The `batch.Count >= 2` is a wrong thing to do--you DO WANT the batches to be as big as `EventHubMessageSizeLimit` allows

Reply:
@karolz-ms I added some additional comments in the code. What I if a batch is too big in total byte size, is to split the batch into two new batches each containing half of the original batch items. And keep doing this until the split-up batch is not too big anymore (and still contains at least 2 items, so it's even possible to split the batch up, hence the `batch.Count >= 2` evaluation.
If you do not appreciate this approach, let me know and I shall do the `foreach` approach.

Reply:
Sorry, I misread the code. It is fine, never mind :-*)